### PR TITLE
Completed T02

### PR DESCRIPTION
**Pull request by Ajay07Metamation**

### CODE REVIEW

**Review by rajeevramanTMM — CHANGES_REQUESTED** [review]

**Comment by rajeevramanTMM** [comment]
Apply changes based on the discussion.

File: Training/Program.cs [hunk 1]
```diff
@@ -1,7 +1,119 @@
...
+      if (Convert.ToString (num).Length == 8 || Convert.ToString (num).Length == 9) {
+         quo = num / 10000000;
+         result += (Convert.ToString (quo).Length == 2 && quo > 20 ? numtoword[quo - (quo % 10)] + " " + numtoword[quo % 10] : numtoword[quo]) + " Crores ";
+         num = num % 10000000;
+      }
+
+   }
+   Write ($"{result}");
+}
+void NumtoRoman (int num) {
```
[the remark this answers is not in the document]
Comment (Ajay07Metamation) [reply Training/Program.cs:62]: Changes done in the latest commit

[the remark this answers is not in the document]
Comment (rajeevramanTMM) [reply Training/Program.cs:62]: Fix this error,
image

[the remark this answers is not in the document]
Comment (Ajay07Metamation) [reply Training/Program.cs:62]: Changes and formatting done in latest commit

File: Training/Program.cs [hunk 2]
```diff
@@ -1,7 +1,75 @@
...
+    { 6, "Six" }, { 7, "Seven" }, { 8, "Eight" }, { 9, "Nine" }, { 10, "Ten" },{11,"Eleven" },{12,"Twelve" },{13,"Thirteen"}
+   ,{14,"Fourteen" },{15,"Fifteen" },{16,"Sixteen" },{17,"Seventeen" },{18,"Eighteen" },{19,"Nineteen" },{20,"Twenty" },
+    {30,"Thirty" },{40,"Fourty" },{50,"Fifty" },{60,"Sixty"},{70,"Seventy" },{80,"Eighty" },{90,"Ninety" } };
+   string result = "", end = "";
+   if (num == 0) WriteLine ("Zero");
+   while (num != 0) {
+      string numstr = num.ToString ();
+      int numlength = numstr.Length;
+      int divident = 0;
+      if (numlength == 1) {
```
Comment (varalakshmi-imm) [inline Training/Program.cs:36]: Whenever you have this kind of multiple if else if then should use switch.
But this code could be simplified without using switch too. 
for eg: numlength == 3, then divident = 100 and if 4 then 1000. So think of simplifying this based on numlength and divident value.

File: Training/Program.cs [hunk 3]
```diff
@@ -1,7 +1,75 @@
...
+      num %= divident;
+   }
+   Write ($"{result}");
+}
+void NumtoRoman (int num) {
+   string[] roman = { "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };
+   int[] number = { 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };
+   string result = "";
+   if (num == 0) WriteLine ("No Roman Number representation for 0");
+   while (num > 0)
```
Comment (varalakshmi-imm) [inline Training/Program.cs:67]: Having brackets will make code clear.

File: Training/Program.cs [hunk 4]
```diff
@@ -1,7 +1,75 @@
...
+      if (numlength == 2) break;
+      num %= divident;
+   }
+   Write ($"{result}");
+}
+void NumtoRoman (int num) {
+   string[] roman = { "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };
+   int[] number = { 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };
+   string result = "";
+   if (num == 0) WriteLine ("No Roman Number representation for 0");
```
Comment (varalakshmi-imm) [inline Training/Program.cs:66]: if (num == 0) it should return after printing the message right?

File: Training/Program.cs [hunk 5]
```diff
@@ -1,7 +1,75 @@
...
+   while (num > 0)
+      for (int i = number.Length - 1; i >= 0; i--)
+         if (num >= number[i]) {
+            result += roman[i];
+            num = num - number[i];
```
Comment (varalakshmi-imm) [inline Training/Program.cs:71]: Should use -= here like previous line +=.
Look for the intellisense underlines and fix them.
image

File: Training/Program.cs [hunk 6]
```diff
@@ -1,7 +1,75 @@
...
-            Console.WriteLine ("Hello, World!");
-        }
-    }
+﻿using static System.Console;
+var key = ConsoleKey.Y;
+while (key == ConsoleKey.Y) {
+   Write ("Enter a Number : ");
+   if (int.TryParse (ReadLine (), out int num) && num >= 0) {
+      Write ("Do you want to convert it into (W)ords or (R)oman Numbers ?  ");
+      switch (ReadKey ().Key) {
```
Comment (varalakshmi-imm) [inline Training/Program.cs:7]: Do you really need this switch case? Just a if else check could do the same job.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 6 above)
Comment (Ajay07Metamation) [reply Training/Program.cs]: All changes mentioned are done in the latest commit mam.

File: Training/Program.cs [hunk 7]
```diff
@@ -1,7 +1,66 @@
...
+      int quo = num / divident;
+      result += (quo.ToString ().Length == 2 && quo > 20 ? numtoword[quo - (quo % 10)] +
+                 " " + numtoword[quo % 10] : numtoword[quo]) + (numlength >= 3 ? endword[numlength] : "") + (numlength == 3 ? "and " : "");
+      num %= divident;
+   }
+   Write ($"{result}");
+}
+
+void NumToRoman (int num) {
+   string[] roman = { "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };
```
Comment (varalakshmi-imm) [inline Training/Program.cs:50]:
```suggestion
   string[] romans = { "I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M" };

```

File: Training/Program.cs [hunk 8]
```diff
@@ -1,7 +1,66 @@
...
+   int[] number = { 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };
```
Comment (varalakshmi-imm) [inline Training/Program.cs:51]:
```suggestion
   int[] numbers = { 1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000 };

```

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
- Write // comments for methods.
- .gitignore is not excluded

File: Training/Program.cs (hunk 8 above)
Comment (Ajay07Metamation) [reply Training/Program.cs]: All changes mentioned are done in the latest commit mam.

File: Training/Program.cs [hunk 9]
```diff
@@ -1,7 +1,79 @@
-﻿namespace Training {
-    internal class Program {
-        static void Main (string[] args) {
-            Console.WriteLine ("Hello, World!");
-        }
+﻿//Numbers to words or roman numbers
```
Comment (varalakshmi-imm) [inline Training/Program.cs:1]:
```suggestion
﻿// Numbers to words or roman numbers
```
Leave a space after //

File: Training/Program.cs [hunk 10]
```diff
@@ -1,7 +1,79 @@
...
-            Console.WriteLine ("Hello, World!");
-        }
+﻿//Numbers to words or roman numbers
+
+using static System.Console;
+
+var key = ConsoleKey.Y;
+while (key == ConsoleKey.Y) {
+    Write ("Enter a Number : ");
+    if (!int.TryParse (ReadLine (), out int num) && num >= 0) {
```
Comment (varalakshmi-imm) [inline Training/Program.cs:8]:
```suggestion
    if (!int.TryParse (ReadLine (), out int num) || num < 0) {

```
You negated the if but not the &&
If you enter the number as -1, your program hangs now.

File: Training/Program.cs [hunk 11]
```diff
@@ -1,7 +1,79 @@
...
+        WriteLine ("Enter valid number");
+        continue;
     }
+    Write ("Do you want to convert it into (W)ords or (R)oman Numbers ?  ");
+    var choice = ReadKey ().Key;
+    if (choice == ConsoleKey.W) {
+        WriteLine ();
```
Comment (varalakshmi-imm) [inline Training/Program.cs:15]: This should be above the line if (choice == ConsoleKey.W) {
        WriteLine ();
image

File: Training/Program.cs [hunk 12]
```diff
@@ -1,7 +1,79 @@
...
+        NumToRomans (num);
+    } else WriteLine ("Enter a valid key");
+    WriteLine ("\nDo you want to convert again (Y)es or (N)o ? ");
+    key = ReadKey ().Key;
+    WriteLine ();
+}
+
+// Method to convert numbers to words using dictionary
+void NumToWords (int num) {
+    Dictionary<int, string> numtowords = new () {
```
Comment (varalakshmi-imm) [inline Training/Program.cs:28]:
```suggestion
    Dictionary<int, string> words = new () {

```

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 12 above)
Comment (Ajay07Metamation) [reply Training/Program.cs]: All changes are done in the latest commit mam

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]
If I enter 10000, the output for Roman conversion is MMMMMMMMMM which is not right.
image
So restrict it to 3999 for both words and romans

**Comment by Ajay07Metamation** [comment]
Done changes

File: Training/Program.cs [hunk 13]
```diff
@@ -1,7 +1,74 @@
...
+        return;
+    }
+    while (num != 0) {
+        int numlength = num.ToString ().Length;
+        int divisor = 0;
+        if (numlength == 1 || (numlength == 2 && num <= 20)) {
+            result += words[num];
+            break;
+        } else if (numlength == 2 && num > 20)
+            divisor = 1;
```
Comment (varalakshmi-imm) [inline Training/Program.cs:44]:
```suggestion
           else if (numlength == 2 && num > 20) divisor = 1;

```
Could be in single line

File: Training/Program.cs [hunk 14]
```diff
@@ -1,7 +1,74 @@
...
+        else
+            divisor = (int)Math.Pow (10, numlength - 1);
```
Comment (varalakshmi-imm) [inline Training/Program.cs:46]:
```suggestion
           else divisor = (int)Math.Pow (10, numlength - 1);

```
Could be in single line

File: Training/Program.cs [hunk 15]
```diff
@@ -1,7 +1,74 @@
...
+        int quo = num / divisor;
+        result += (quo.ToString ().Length == 2 && quo > 20 ? words[quo - (quo % 10)] + " " + words[quo % 10] : words[quo])
```
Comment (varalakshmi-imm) [inline Training/Program.cs:48]: See if you can split each plus part in this line into meaningful variables and finally merge into one.  
Also quo % 10 is done twice in this statement. Could have stored in a variable and accessed it.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 15 above)
Comment (Ajay07Metamation) [reply Training/Program.cs]: All changes done in the latest commit mam

File: Training/Program.cs [hunk 16]
```diff
@@ -1,7 +1,75 @@
...
+        int numlength = num.ToString ().Length;
+        int divisor = 0;
+        if (numlength == 1 || (numlength == 2 && num <= 20)) {
+            result += words[num];
+            break;
+        } else if (numlength == 2 && num > 20) divisor = 1;
+        else divisor = (int)Math.Pow (10, numlength - 1);
+        int quo = num / divisor;
+        var unitDigit = quo % 10;
+        var formattedQuo = quo.ToString ().Length == 2 && quo > 20 ? words[quo - unitDigit] + " " + words[unitDigit] : words[quo];
```
Comment (varalakshmi-imm) [inline Training/Program.cs:48]: Not so good name formattedQuo.
But I too can't think of any suggestions.
So let it be.

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 16 above)
Comment (Ajay07Metamation) [reply Training/Program.cs:47]: All changes done mam

File: Training/Program.cs [hunk 17]
```diff
@@ -1,7 +1,74 @@
...
-        static void Main (string[] args) {
-            Console.WriteLine ("Hello, World!");
-        }
+﻿// Numbers to words or roman numbers
+
+using static System.Console;
+
+var key = ConsoleKey.Y;
+while (key == ConsoleKey.Y) {
+    Write ("Enter a Number between 0 and 3999 : ");
```
Comment (varalakshmi-imm) [inline Training/Program.cs:7]: Still haven't formatted this program.
Click Edit > Advanced > Format Document.
image

**Review by varalakshmi-imm — CHANGES_REQUESTED** [review]

File: Training/Program.cs (hunk 17 above)
Comment (Ajay07Metamation) [reply Training/Program.cs]: Mam I have done it again and pushed it mam

**Review by varalakshmi-imm — APPROVED** [review]

File: Training/Program.cs [hunk 18]
```diff
@@ -1,7 +1,74 @@
...
+      int len = num.ToString ().Length;
+      int divisor;
+      if (len == 1 || (len == 2 && num <= 20)) {
+         result += words[num];
+         break;
+      } else if (len == 2 && num > 20) divisor = 1;
+      else divisor = (int)Math.Pow (10, len - 1);
+      int quo = num / divisor;
+      var unitDigit = quo % 10;
+      var formattedQuo = quo.ToString ().Length == 2 && quo > 20 ? words[quo - unitDigit] + " " + words[unitDigit] : words[quo];
```
Comment (varalakshmi-imm) [inline Training/Program.cs:47]: @Ajay07Metamation This has become long line after formatting.
Need to watch such lines after formatting.

**Review by varalakshmi-imm — APPROVED** [review]

**Pull request closed without merging**